### PR TITLE
nixos/nginx: add option typesHashMaxSize

### PR DESCRIPTION
**Pull request by Izorkin**
## Description of changes
Allow to change maximum size of the types hash tables.
It is recommended to reduce the size if a custom file with mime-types is used.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pbsds** [comment]
>  It is recommended to reduce the size if a custom file with mime-types is used.

source?

This strikes me like the kinda thing that explodes at runtime, not build/eval time. Can this be added to the tests?

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (pbsds) [inline nixos/modules/services/web-servers/nginx/default.nix:899]: image


```suggestion
        defaultText = literalExpression ''if cfg.defaultMimeTypes == "''${pkgs.mailcap}/etc/nginx/mime.types" then 2688 else 1024'';
```

Comment (pbsds) [reply nixos/modules/services/web-servers/nginx/default.nix:899]: also
```suggestion
        defaultText = literalExpression "if config.services.nginx.defaultMimeTypes == ''$''{pkgs.mailcap}/etc/nginx/mime.types'' then 2688 else 1024";
```

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:898]: I think we should move this into a let in to prevent it accidentally getting out of date.

**Comment by Izorkin** [comment]
> source?

https://nginx.org/en/docs/hash.html

> This strikes me like the kinda thing that explodes at runtime, not build/eval time. Can this be added to the tests?

I'll see what I can do.

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix]: Thanks, fixed.

**Comment by Izorkin** [comment]
Updated test.

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:898]: The mime.types file is rarely updated. If it is, the test will show that a change will be required.

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (pbsds) [inline nixos/modules/services/web-servers/nginx/default.nix:908]:
```suggestion
          Sets the maximum size of the types hash tables (`types_hash_max_size`).
          It is recommended that the minimum size possible size is used.

          If {option}`recommendedOptimisation` is disabled, nginx would otherwise
          fail to start since the mailmap `mime.types` database has more entries
          than the nginx default value 1024.
```

**Review by pbsds — COMMENTED** [review]
I played a bit around with this.
If set to 1024:
```
vm-test-run-nginx> subtest: check optimal size of types_hash
vm-test-run-nginx> webserver: must fail: journalctl --unit nginx --grep 'could not build optimal types_hash'
vm-test-run-nginx> Test "check optimal size of types_hash" failed with error: "command `journalctl --unit nginx --grep 'could not build optimal types_hash'` unexpectedly succeeded"
```

if set to 4096, it passes.

if set to 16400, it passes.


How did you arrive at `2688`? Is there any way to test for the max_size being too big? (assuming there is a performance penalty for that)

File: nixos/modules/services/web-servers/nginx/default.nix
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix]: Thanks, fixed.

**Comment by Izorkin** [comment]
> How did you arrive at `2688`? Is there any way to test for the max_size being too big? (assuming there is a performance penalty for that)

I adjusted the parameter manually, in increments of 64 and the current `mime.types` file in the `mailcap` package.
If use a smaller mime.types file, the types_hash_max_size can be reduced.

**Review by pbsds — APPROVED** [review]
LGTM, i leave it to code-owner to decide if the `typesHashMaxSize` limit should be this tight, and whether the `types_hash` test should be a (i assume) release blocker.

**Comment by Izorkin** [comment]
cc @RaitoBezarius @pbsds

**Review by fpletz — APPROVED** [review]
LGTM. I think we don't need the nginx test to be a release blocker since this technically doesn't render nginx to be broken. It should only incur a small performance hit. The failing test should be enough to cover this.

**Comment by Izorkin** [comment]
@fpletz is that an variant?

**Comment by fpletz** [comment]
> @fpletz is that an variant?

Sorry, I didn't mean any change by my review. I'm fine with either the nginx test standalone or having the nginx test in mailcap. My only intention with the review was to convery that the nginx test doesn't need to be a release blocker - if anybody feels otherwise I'm also fine with that.

**Comment by Izorkin** [comment]
Updated and rebased PR.

**Review by fpletz — APPROVED** [review]

**Pull request merged**
